design/healcheck-endpoint.md

```diff
@@ -0,0 +1,47 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check (see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)). Currently for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL. This approach has many drawbacks, here are a few examples:
+
+* We can not discover database failures
+* We can not discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address this problems, we should provide a parametrized Healthcheck endpoint in the Server and use it in our container image as well as in the Operator. 
+
+## Parametrizing the Healthcheck URL
+
+The preferable approach is to hook into Health Endpoint provided by Wildfly (starting from 16) at:
+
+    http://localhost:9990/health
+
+The endpoint is provided according to the [MicroProfile Health](https://microprofile.io/project/eclipse/microprofile-health) specification.
+
+The implementation should take additional parameters into consideration, including:
+
+* `timeout` - the request gets terminated if not performed within a timeout. The implementation needs to take special care of releasing the thread pool (if necessary). Subsequent calls that take longer than expected should not exhaust the thread pools inside Wildfly.
+* `fail-on` - a set of conditions under which the probe fails. The initial implementation should cover:
+  * `db` - failing on database failures
+  * `idp-<prefix>` - failing when an IdP is not accessible
+  * `ldap` - failing is LDAP connection is down
```

Comment:
LDAP is same as IdPs. There may be multiple LDAP instances, as well as other user federation providers. So should probably be "user-federation-<id>"

Reply:
I totally agree. I'll add this.

Reply:
Addressed in the next iteration. Resolving.